### PR TITLE
My code is not passing tests and I'm stuck on the listIterator

### PR DESCRIPTION
**Pull request by TBarrett553**
I'm working on the poorly made conditionals, right now. I apologize this should have been finished last week. I would really appreciate your input. My main problem is that the listIterator is infinite and it only works if I put a break at the end.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,96 @@
...
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+    final int length = (int) request.getDuration();
+    Collection<String>  requestAttendees = request.getAttendees();
+    
+    // Create a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Check that there are booked meetings before sorting it and running the available function
+    if (bookTimes.isEmpty()) {
+            availTimes.add(bookTimes.get(0).WHOLE_DAY);
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Hmm, how does this work? Won't bookTimes.get(0) throw if bookTimes.isEmpty()?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,96 @@
...
+    // Create a List of relevant events that will restrict available time ranges
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+
+    for (Iterator<Event> eventIterator = events.iterator(); eventIterator.hasNext();){
+        Event event = eventIterator.next();
+      
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee) && !bookTimes.contains(event.getWhen())){
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If you used a HashSet for bookTimes, then you wouldn't have to check bookTimes.contains(event.getWhen()). And it would be more efficient.

https://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (flying-bus) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If you're just trying to access the static variable, then I think you can just say: TimeRange.WHOLE_DAY.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,96 @@
...
+    // Create a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Check that there are booked meetings before sorting it and running the available function
+    if (bookTimes.isEmpty()) {
+            availTimes.add(bookTimes.get(0).WHOLE_DAY);
+    }
+
+    else {
+        Collections.sort(bookTimes, bookTimes.get(0).ORDER_BY_START);
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Static variable just use the classname:
TimeRange.ORDER_BY_START

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,96 @@
...
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+    return bookTimes;
+  }
+
+  public Collection<TimeRange> availableTimes(ArrayList<TimeRange> bookTimes, int length){
+    Collection<TimeRange> availTimes = new HashSet<>();
+    if(length > bookTimes.get(0).END_OF_DAY) {
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'm not sure what this means - can you add a comment explaining what it does?

If the length of the event is more than a day, then there are no available times?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,96 @@
...
+    
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(bookIterator.hasPrevious()) {
+                TimeRange prevBookedTime = bookIterator.previous();
+                if(!bookedTime.overlaps(prevBookedTime) && (prevBookedTime.end() + length <= bookedTime.start())) {
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.start(), false));
+                }
+
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: A good way to format if/else is like this:

if (cond) {

} else if (cond) {

} else {

}

That way the formatting gives an extra clue that you're looking at an else/if condition.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,96 @@
...
+        return availTimes;
+    }
+// My problems are in the for loop.I believe that my conditionals are inefficient and I'm still working on it.
+    else {
+    
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(bookIterator.hasPrevious()) {
+                TimeRange prevBookedTime = bookIterator.previous();
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Ah ha, I think I found your bug.

When you call "previous" it actually moves the iterator backwards! So on Line 77, you move the iterator forwards, and then Line 80 moves it backwards again, so you never actually manage to advance the iterator.

If you want to actually store the previous TimeRange, then this might work:

```
TimeRange prevBookedTime = null;
for (ListIIterator<TimeRange> bookIterator = /* like before */) {
  TimeRange bookedTime = bookIterator.next();
  if (prevBookedTime != null) {  // instead of bookIterator.hasPrevious()
     if (!bookedTime.overlaps(prevBookedTime) /* ... like before ... (/ ) {
     }
  }
 
  // Store bookedTime in prevBookedTime for the next iteration through the loop.
  prevBookedTime = bookedTime;
}
```

https://docs.oracle.com/javase/7/docs/api/java/util/ListIterator.html#previous()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,96 @@
...
+                if(!bookedTime.overlaps(prevBookedTime) && (prevBookedTime.end() + length <= bookedTime.start())) {
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.start(), false));
+                }
+
+                else if(bookedTime.overlaps(prevBookedTime)) {
+                    if(prevBookedTime.contains(bookedTime)){
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.END_OF_DAY, true));
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Use TimeRange.END_OF_DAY and TimeRange.START_OF_DAY here and below.

**Review by flying-bus — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: There is a test that requires that the whole day should be returned if the day is open. In other words, if the bookTimes is empty I wanted to skip the whole list iterator function. There is probably a more elegant way to do this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Wow thank you! When I kept looking into this error I couldn't find an example when .previous() and .next() were used in the same iterator instance. But this makes sense, pointers are a tricky thing I'm still learning about.

**Comment by freeboson** [comment]
I'll take a look after you make the changes @flying-bus requested. Ping me here or in dynamite when you're ready.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The main reason I used an ArrayList was so I could sort it by time. But I initially made a HashSet to go through the for loop and then converted it to a list so it can be sorted.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thank you and makes a lot of sense!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This was in consideration of a test (too long of a duration). The test says that the code should return an empty list if the requested duration is longer than a whole day.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: I tried this formatting in my recent commits. Was it done correctly?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Got it!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,97 @@
...
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatizes return data structure and set duration parameter
+    //Collection<TimeRange> availTimes = new HashSet<>();
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: rm

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,97 @@
...
+
+  public Collection<TimeRange> getAvailableTimeRanges(ArrayList<TimeRange> bookTimes, int length){
+    ArrayList<TimeRange> availTimes = new ArrayList<TimeRange>();
+
+    // There is a test that doesn't allow a duration longer than the day. The proper result is to return an empty list
+    if(length > TimeRange.END_OF_DAY) {
+        return availTimes;
+
+    // If there are no Time Conflicts the whole day should be returned as available
+    } else if (bookTimes.isEmpty()) {
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This doesn't need an else

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,97 @@
...
+        availTimes.add(TimeRange.WHOLE_DAY);
+        return availTimes;
+
+    } else {
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This shouldn't be an else

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,97 @@
...
+    // Even though previous() is not used ListIterator must be used with iterating through ArrayList
+        TimeRange prevBookedTime = null;
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(prevBookedTime != null) {
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Need a comment before this with what you're doing here. E.g.,

// If we've already found a booked time, extend it

(if that's what the behavior is -- i can't really tell)

**Review by freeboson — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -19,91 +19,92 @@
...
-            if(event.getAttendees().contains(attendee) && !bookTimes.contains(event.getWhen())){
+            if(event.getAttendees().contains(attendee)){
                 bookTimes.add(event.getWhen());
             }
         }
     }
-    return bookTimes;
+
+    // The times have to be returned as a list to be sorted by start time
+    ArrayList<TimeRange> bookedTimes = new ArrayList<TimeRange>(bookTimes) ;
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: The next two lines may be simpler to do:

return new ArrayList<TimeRange>(bookTimes);

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (flying-bus) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: The formatting LGTM (Looks Good To Me)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -19,91 +19,92 @@
...
+            if(prevBookedTime != null) { 
+                if(!bookedTime.overlaps(prevBookedTime) && prevBookedTime.end() + length <= bookedTime.start()) {
                     availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.start(), false));
-                }
 
-                else if(bookedTime.overlaps(prevBookedTime)) {
-                    if(prevBookedTime.contains(bookedTime)){
-                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.END_OF_DAY, true));
-                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.START_OF_DAY, bookedTime.END_OF_DAY, true));
+                    if(!bookIterator.hasNext() && bookedTime.end() + length <= TimeRange.END_OF_DAY) {
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: There are a lot of nested if-statements here, which makes it sort of hard to follow. Some of them look very similar? (Like here and Line 92).

Is it possible to simplify the conditions a bit? Or if not, can the repeated code be put into a helper function? Or at the very least, add some comments which state what the conditions are.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
...
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeeConsidered() {
+      // Have each person have different events. We should see three options because each person has
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:310]: s/three/two/

(this is common notation for replace "three" with "two" based syntax for the tool "sed").

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: Now removed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: So basically I should make it a separate if statement?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 15]
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
...
+    // split the restricted times. There is an optional attendee and their event is considered.
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:311]: The previous test doesn't consider the optional attendee, but this test does. I don't obviously see where that option is set (is it a flag? or something else that triggers it?).

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 16]
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void optionalAttendeesWithoutGaps() {
+    // There are two optional meeting attendees. They have several gaps in their schedule that
```
Comment (flying-bus) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Is this comment accurate?

**Review by flying-bus — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (freeboson) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yes, because there's an early return (good — prefer early return over nested ifs)

```
if (cond1) {
  return ___;
}
if (cond2) {
  return ___;
}
```

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 15 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:311]: It seems that the optional attendee's schedule is considered if their event has the same TimeRange as an available meeting time. It's kind of an odd requirement but that's why the optional attendee is considered in the test.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Now I see that it's unnecessary

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Great thank you for that simplification!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (freeboson) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: suggestion from 1:1 preprocess `bookTimes` to be disjoint/deduped.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,115 @@
...
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+
+    // The times have to be returned as a list to be sorted by start time 
+    return new ArrayList(bookTimes);
+  }
+
+  public ArrayList<TimeRange> collapseTimeConflicts (ArrayList<TimeRange> bookTimes){
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This method both mutates its argument, and returns it. Consider making it either:

```
public List<TimeRange> collapseTimeConflicts(ImmutableList<TimeRange> bookTimes) {
  // ...
  return value;
}
```

or:

```
public void collapseTimeConflicts(List<TimeRange> bookTimes) {
  // mutate bookTimes directly
}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,115 @@
...
+
+        TimeRange prevBookedTime = null;
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();){
+            TimeRange bookedTime = bookIterator.next();
+            if(prevBookedTime != null){
+                // For nested events remove the inner event and keep the longer event Time Range
+                if(prevBookedTime.contains(bookedTime)){
+                    bookTimes.remove(bookedTime); 
+
+                // For overlapping events create a new Time Range with the earliest start time and latest end time
```
Comment (freeboson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:78]: It's worth noting that the case where bookedTime.end() is earlier than prevBookedTime.end() is handled by the `.contains()` case above.

**Review by freeboson — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 15 above)
Comment (flying-bus) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:311]: I think I just don't see where in the code the optional attendees are considered at all.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I was able to implement the disjoint method.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah that was a mess I was able to clean it up and I placed new comments with the refigured algorithm.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 17 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: So essentially, if I mutate the list I don't need to return it?

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 14 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:310]: Got it! Very interesting

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 15 above)
Comment (TBarrett553) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:311]: I don't have the optional feature in my code because the walk through stated that if there is extra time I could add code for the optional tests. Unfortunately I was behind, so I couldn't prioritize this section.

Comment (flying-bus) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:311]: ok, sgtm.

**Review by flying-bus — APPROVED** [review]
